chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,84 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T<..>).
+    /// Implemented(!T<..>: A).
+    /// Implemented(!T<..>: B).
+    /// Implemented(!T<..>: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T<..>` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        let interner = builder.interner();
+        let substitution = builder.substitution_in_scope();
+        let alias = AliasTy::Opaque(OpaqueTy {
+            opaque_ty_id: self.opaque_ty_id,
+            substitution: substitution.clone(),
+        });
+
+        let alias_ty = Ty::new(
+            interner,
+            ApplicationTy {
+                name: TypeName::OpaqueType(self.opaque_ty_id),
+                substitution,
+            },
+        );
+
+        builder.push_binders(&self.bound, |builder, opaque_ty_bound| {
+            // AliasEq(T<..> = HiddenTy) :- Reveal.
+            builder.push_clause(
+                DomainGoal::Holds(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: opaque_ty_bound.hidden_ty.clone(),
+                    }
+                    .cast(interner),
+                ),
+                iter::once(DomainGoal::Reveal(())),
+            );
+
+            // AliasEq(T<..> = !T<..>).
+            builder.push_fact(DomainGoal::Holds(
+                AliasEq {
+                    alias: alias.clone(),
+                    ty: alias_ty.clone(),
+                }
+                .cast(interner),
+            ));
+
+            for bound in &opaque_ty_bound.bounds {
+                // Implemented(!T<..>: Bound).
+                builder.push_binders(&bound, |builder, bound| {
```

Comment:
This I think is not quite right either. These binders represent the hidden self-type, right? I think they want to be *instantiated* with the placeholder. That is, right now given (say) `type Foo<P1..Pn> = impl Trait<T1..Tn>` you're generating something like

```
forall<P1..PN> { // generated by the first push_binders
    forall<X> { // generated by the second push_binders
        forall<> { // generated by the 3rd push_binders; we don't have `for<'a>` or anything, so empty
            X: Trait
        }
    }
}
```

instead I think what you want is to do 

```rust
for bound in &opaque_ty_bound.bounds {
  // Implemented(!T<..>: Bound).
  let bound_with_placeholder_ty = bound.substitute(interner, &[alias_placeholder_ty]);
  builder.push_binders(&bound_with_placeholder_ty , |builder, bound| {
    builder.push_fact(bound.into_well_formed_goal(interner));
  });
}
```